Change the Order of Cancellation API Calls to Meet Requirements

dzver opened the pull request:
Part of DOTCOM-13647

## Proposed Changes

For WordPress.com:

* Redesign the first step of the cancellation flow to merge the first/last step in Atomic plans and Domains step
* Cancel and refund after clicking the Red button
* Introduce a loading state of the Red button
* Show success notice right above the cancellation survey (when a cancellation survey is available)
* Update all buttons and copy indicating that the cancellation hasn't happened with copy that's more neutral (submit, cancel, next)
* Update all upsell steps to point to checkout
* Back buttons should point to me/purchases

For WordPress.com domains (new change per request):
* Introduces one-step full cancellation behind a checkbox for domains that have a refund
* Domains that do not have a refund would have their auto-renew turned off
* Domains that are already auto-renewed keep an ugly modal with a blue button (not subject to this change)
* Post-cancellation survey introduced (won't be visible in MC, needs a separate change)

<img width="48%" alt="Screenshot 2025-06-27 at 17 13 57" src="https://github.com/user-attachments/assets/8cf86402-d6cb-4698-b274-fd9f130ff257" /> <img width="48%" alt="Screenshot 2025-06-27 at 17 14 12" src="https://github.com/user-attachments/assets/decb1dba-3eb6-481c-9bdb-afb04a829ed2" /> 

For Jetpack:

* Hide for now the first step of the cancellation flow that shows that you have 11 months remaining
* Trigger the cancellation API immediately after the cancel is button is clicked
* Edge cases when the cancellation is outside of refund period or paid with credits have been addressed

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

*

## Testing Instructions


You'll need to try this with at least a Jetpack product and a WordPress.com product, and the different products have different cancellation flows. Here are videos of the main ones:

For WPCOM:
* Simple:
* With and without domain
* Atomic
* With and without domain
* Try Space or Email (same flow)

Both cases have a sub-case when the plan is not refundable

A video of the most complicated flow, with-refund, Atomic, and domains. Note that the checkboxes become disabled during the cancellation and the back button post-refund goes to /me/purchases

https://github.com/user-attachments/assets/68d2d9dc-6e89-4227-a09a-db5956a70ead

For Jetpack, try Boost, Social

Video of the most common case:
https://github.com/user-attachments/assets/c70a76c0-19f3-4ae8-abf0-476ae7424775

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

dzver committed:
Temporary commit

dzver committed:
Work in progress - temporary commit

dzver committed:
Replace divs with notices

dzver committed:
Show dismiss to the notice

dzver committed:
Style the notice to be roughly as everything else in Calypso

dzver committed:
Temporary commit:

I want to remove these but not there yet

dzver committed:
Remove redundant variable leftover from previous attempts

dzver committed:
This is also an interim solution

dzver committed:
Rename class names

dzver committed:
Implement a relevant cancellationFailed function

dzver committed:
Adjust the success message

dzver committed:
Fix missing props

dzver committed:
Restore the call to remove marketplace subscriptions

This is no longer in submitCancelAndRefundPurchase as it now returns
value

dzver committed:
Add Atomic Revert changes similar to the plan features in cancellation

dzver committed:
Use a caution gridicon

dzver committed:
Add gridicon

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17571005

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @dzver for including a screenshot in the description! This is really helpful for our translators.

dzver committed:
Update Atomic Revert Changes component, add checkbox

dzver committed:
Styling and TS error

dzver committed:
Additional copy changes for the Atomic revert step

dzver committed:
The Atomic revert confirmation is only needed if it's not a new site

This way there will be no two overlapping confirmations

dzver committed:
Styling fixes

fditrapani committed:
CSS tweaks

fditrapani committed:
Reorder content

fditrapani committed:
Hide title for non-atomic plans

dzver committed:
Point downgrade solutions to checkout rather than API

dzver committed:
Update the submit button

dzver committed:
Restrict the free month offer to non-refundable monthly plans

dzver committed:
Update the refund success message

dzver committed:
Jetpack: display notice

dzver committed:
Remove Jetpack offer - it no longer works

dzver committed:
Restore cancel message for Jetpack

dzver committed:
Fix the different checkbox styling

It comes from using different components

dzver committed:
Undo changes to yarn lock

These slipped in with design changes early on and I missed it

dzver committed:
Revert interim CSS

dzver committed:
Let's not accidentaly disable this button for storage add-ons

dzver committed:
Shorten the message and make it more universal

dzver committed:
A less radical removal of the atomic cancel confirmation

It's still needed for removals

dzver committed:
Standartize post-survey notices

fditrapani committed:
Add title and tweak checkbox spacing

fditrapani commented:
Thanks for the updates @dzver! This is mostly in a good spot. I found a few things and I pushed some minor updates. Some CSS tweaks and adding that title back when the domain options are visible. We need it to separate the domain options for the content.

<img width="1840" alt="image" src="https://github.com/user-attachments/assets/8e4dc127-96d2-4d3c-bde3-3946345091d8" />


This is what I found during testing:

1. There's a delay for the confirmation to appear after you press the cancel button. It's a lot more pronounced when you process a refund or you do a jetpack product. For Jetpack it does this until the copy loads:

<img width="1837" alt="image" src="https://github.com/user-attachments/assets/1e69f100-7dc9-473b-a2b9-3dbaa3bc28a1" />


The notice can be easily missed and it makes the content jump when it appears. Can we keep people on the confirmation screen while we process the cancellation and then take them to the survey after it's done? That way we can show the notice as soon as you land there.

We'd have to disable all the form elements on the page and set the cancel button to is-active?

<img width="1840" alt="image" src="https://github.com/user-attachments/assets/3a96a230-fd68-49a9-8d34-165e979b4eff" />

4. If I process a refund, then press back from the survey screen, it takes me back to the confirmation screen with this error:

<img width="1840" alt="image" src="https://github.com/user-attachments/assets/cff3467d-4a6b-4838-8a5c-3850d587866f" />

Can we make the back button take you to your subscriptions (since the produce has been removed at this point).

5. I tested cancelling a plugin, Google mail, a storage add-on, and professional mail. They both worked as expected. Similar to #1, it takes longer to process these so it feels like an eternity waiting for the confirmation notice to appear.

6. When you cancel a domain, we try understand your intent after you cancel. I think this step makes more sense before you get the cancellation confirmation. We should also change the CTA to "Continue with cancellation" and make it red too.

<img width="1115" alt="image" src="https://github.com/user-attachments/assets/61e51d05-2a74-48bb-a8ca-56f0acac452f" />

7. After you complete the survey, we take you to the subscription list and show you another notice that the subscription has been cancelled. Can we remove the notice and for an extra bonus take people back to the purchase settings screen if it hasn't been removed yet?

dzver committed:
Remove space

dzver committed:
Address the empty notice

dzver committed:
Make this notice impossible to close

dzver committed:
Filippo's request for loading state on the cancel button

dzver committed:
Ensure people can't go back

dzver committed:
Remove another step per request

Jetpack features step, which usually shows that you have 11 months
remaining of your 12 month subscriptions minutes after purchase, is
being hidden.

This change won't do a code cleanup

dzver committed:
Unify buttons for domains per Filippo's request

fditrapani commented:
Awesome! thank you. I'll take a look a little later. No problem on doing number 7 later.

jboland88 commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -395,12 +274,14 @@ const CancelJetpackForm: React.FC< Props > = ( {
 					<FormattedHeader
-						headerText={ translate( 'Confirm removal' ) }
+						brandFont
+						headerText={ translate( 'Sorry to see you go' ) }
```
On the Jetpack flow, the text indicates to click "Remove Subscription", but the button says "submit" 

![Screenshot 2025-06-26 at 11 08 25 AM](https://github.com/user-attachments/assets/76aafb5c-d29e-4c88-a797-cc52f96490e0)

fditrapani replied:
This looks like one of those duplicate cases confirming the cancellation after confirming it already. Do we still need this screen?

jboland88 replied:
This one actually pops up after clicking "Remove subscription" from the purchase management screen for Jetpack purchases, so in this case, it's a single confirmation.

![Screenshot 2025-06-26 at 4 44 30 PM](https://github.com/user-attachments/assets/b3869d33-4e5e-4216-aa64-5c2cc2964b54)

fditrapani replied:
Weird. In that case, we'll want to use the default cancellation confirmation screen. Something like these:

![image](https://github.com/user-attachments/assets/d36315a2-f4a6-45c2-a1fb-135a24aa14e6)
![image](https://github.com/user-attachments/assets/4b3d2801-534b-4c7b-8b27-dfdeb527cb0d)

dzver replied:
It pops after Remove subscription on an expired plan but not after clicking Remove subscription on a regular plan. Something to keep in mind when we work on unifying the flows

jboland88 commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -308,80 +220,47 @@ const CancelJetpackForm: React.FC< Props > = ( {
 				busy={ disabled }
 				onClick={ onSubmit }
 				primary
 				scary
```
When showing on the survey as the last step, it feels odd to have the "Submit" button be the red color since there's no destructive action happening

![Screenshot 2025-06-26 at 11 24 31 AM](https://github.com/user-attachments/assets/113e69d2-37cd-408a-ba86-fb4727c25210)

fditrapani replied:
Good call out! This should be a regular button and not a scary one.

dzver replied:
I think I resolved that as well - blue CTA + 'Submit' if the cancellation is processed, red + 'Cancel subscription' if not. Not sure why we have two different starting points for the flow but maybe we can unify these later on as well.

jboland88 commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -60,17 +56,8 @@ const CancelJetpackForm: React.FC< Props > = ( {
-			return steps.FEATURES_LOST_STEP;
-		}
-
 		return steps.CANCELLATION_REASON_STEP;
```
Removing this condition is causing both the list of features lost and the cancellation survey to show when the flow type is "remove" when we only want to show the features lost for the removal flow.

I tested this with a purchase with a payment method that was past the refund window.
- On cancel I got: Survey
- On remove I got: Survey, list of features lost

IIRC the intent was to show the cancellation survey only when the subscription was cancelled and then show the benefits lost only before the subscription was removed to warn the user of the features that would be disabled.

dzver replied:
Yeah, nice catch!

The list of features lost was the default case, which was hit due to a step being unavailable. I restored the deleted code and added a few ifs so we now see it when the subscription is outside of the refund period and about to be removed.

<img width="1211" alt="Screenshot 2025-06-27 at 11 01 04" src="https://github.com/user-attachments/assets/6acea1ff-8cee-469e-b6a9-752ce9f0c101" />

jboland88 commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -365,17 +392,20 @@ class CancelPurchaseButton extends Component {
 						onClose={ this.closeDialog }
-						onClickFinalConfirm={ this.submitCancelAndRefundPurchase }
+						onSurveyComplete={ this.handleSurveyComplete }
 						flowType={ getPurchaseCancellationFlowType( purchase ) }
```
There's an odd bug happening with Jetpack flows and purchases made with credits/ missing payment method still in the refund window that I believe has to do with how we assess the flow type here.

Because there is no refundable amount for purchases with credits, the flow type gets evaluated as "REMOVE" here, and the step with the list of benefits shows asking if the user is sure they want to remove the subscription.

So first, when I go to cancel a subscription made with credits - I see the language that makes it look like I'm going to get a refund and the purchase will get removed right away - but there's nothing to refund.
However, the effect of cancelling is just to disable the auto-renew parameter (which for some reason is set to 1 for a purchase made with no payment method?)
![Screenshot 2025-06-26 at 11 52 13 AM](https://github.com/user-attachments/assets/e101ef50-0a42-44c3-b03b-c201e5e3ba39)

After clicking "cancel" I get the survey and a confirmation that the subscription is still usable and not removed.
![Screenshot 2025-06-26 at 11 52 32 AM](https://github.com/user-attachments/assets/eae113a3-cadb-4c0c-958b-0ae91349e0bf)

But I also get the step warning me about what I will lose if the subscription is removed. This makes it sound like the subscription is getting removed
![Screenshot 2025-06-26 at 11 52 39 AM](https://github.com/user-attachments/assets/fdd97a87-345d-44ba-80fc-19cd8d412db2)

The confirmation message indicates that the subscription was just cancelled, not removed
![Screenshot 2025-06-26 at 11 52 45 AM](https://github.com/user-attachments/assets/0d9a9991-db91-47d9-ba41-4c05eec1b9e7)

fditrapani replied:
> Because there is no refundable amount for purchases with credits, the flow type gets evaluated as "REMOVE" here, and the step with the list of benefits shows asking if the user is sure they want to remove the subscription.

Good catch. This aligns with an upcoming piece of work to separate cancelling a subscription + removing/refunding. Right now, you don't have an option to do anything but remove/refund if you're in the refund window. Since this is the current behaviour, I think we can leave as is and follow up in our next project to separate cancelling + refunding.

> After clicking "cancel" I get the survey and a confirmation that the subscription is still usable and not removed.

That's weird. It's not consistently understanding what's happening here.

> But I also get the step warning me about what I will lose if the subscription is removed. This makes it sound like the subscription is getting removed

ugg. This is one of those duplicate confirmation screens we need to remove now. We have the calypso cancellation confirmation screen that shows before this. This shouldn't be necessary. I guess it never was.

dzver replied:
Half of the problem got resolved by previous changes.

The other half is a rare production bug. Documented it here DOTCOM-13716

jboland88 reviewed (commented):
I've just done some Jetpack testing so far and noted a few things. I'll come back to test the WordPress.com flows

dzver commented:
> Also, I don't know if it's just me but I'm not seeing the domain options anymore. Did something change in the logic for that?

It needs a bundled domain? Perhaps the domain no longer bundled?

<img width="1380" alt="Screenshot 2025-06-26 at 22 33 39" src="https://github.com/user-attachments/assets/3e1487bf-2161-4fd3-99a3-e2e4b53f3962" />

Overall, some very good finds here! So far nothing scary (except the button that's scary and shouldn't be)

jboland88 commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -86,97 +132,66 @@ class CancelPurchaseButton extends Component {
+				return {
+					success: true,
+					message: translate(
+						'%(purchaseName)s was successfully cancelled. It will be available for use until it expires on %(subscriptionEndDate)s.',
```
It could be nice to include the site url here like we do in the removal message. Not sure if that makes it too long? If I have multiple purchases/ sites it could be a nice confirmation that, yes, I removed the right one.

![Screenshot 2025-06-26 at 4 51 59 PM](https://github.com/user-attachments/assets/b9a8a543-7492-428d-84de-2e748a5c2963)

fditrapani replied:
This notice is quite long and it duplicates the messaging on the previous screen. We have a linear issue to update the behaviour so that we redirect back to the purchase sessions which would reinforce the purchase/site we just updated as you're suggestion.

dzver replied:
I can see that it would be very nice for testing the cancellation flow, to get a reminder about what was just cancelled, something that's relevant to my test users with multiple sites on different plans. I think it's out of scope for this change.

jboland88 commented on client/components/marketing-survey/cancel-purchase-form/index.jsx:
```diff
@@ -751,7 +732,21 @@ class CancelPurchaseForm extends Component {
+									</Notice>
+								</div>
+							) }
+							{ this.surveyContent() }
```
Adding the comment here because the deeper survey content is not part of this PR
So, this is a note for a future and separate PR:

Would it make sense to add a small link to "skip survey" or "back to purchases" below the submit button when filling the survey?
I know we've got the back button and that the clear indicator the subscription is already cancelled, but something about clicking "Back" in the top right feels like I might undo the cancellation? Just a thought.

Alternatively, instead of just "Back" in the top left, maybe that could say "back to purchases"?

![Screenshot 2025-06-26 at 4 49 57 PM](https://github.com/user-attachments/assets/19886e98-1fe0-419e-aa56-e6eb187655ef)

fditrapani replied:
Good suggestion! We have plans to add a skip button in the new design:

![image](https://github.com/user-attachments/assets/7c2874a1-65ca-4249-9dea-c37142cb3aa4)

We have a linear task created to update the survey so we can add it then.

jboland88 commented on client/me/purchases/cancel-purchase/index.jsx:
```diff
@@ -349,13 +404,6 @@ class CancelPurchase extends Component {
 							cancellationFeatures={ cancellationFeatures }
```
Note for future iterations @fditrapani - I think it would be great to get Jetpack moved to use this cancellation features list display instead of showing the features that will be lost in a modal.

One thing that I like about what Jetpack does though is showing the features lost before removal - which is helpful to show what's about to stop working.

![Screenshot 2025-06-26 at 5 12 05 PM](https://github.com/user-attachments/assets/004731f1-eb5a-479b-82d9-f88ffe1b006b)

fditrapani replied:
Absolutely! All good points. The new design will present the features on purchase settings screen, before the cancellation and to hopefully prevent people from pressing the cancel in the first place:
![image](https://github.com/user-attachments/assets/8139c514-fe2a-4b66-9b0b-674da9728c41)

The we'll echo the features they'll lose on the cancel screen.
![image](https://github.com/user-attachments/assets/06c11a41-6781-45f5-ad4f-dddca72dce34)

jboland88 commented on client/me/purchases/cancel-purchase/button.jsx:
```diff
@@ -349,12 +373,15 @@ class CancelPurchaseButton extends Component {
 						isVisible={ this.state.showDialog }
 						onClose={ this.closeDialog }
-						onClickFinalConfirm={ this.submitCancelAndRefundPurchase }
+						onSurveyComplete={ this.handleSurveyComplete }
```
I am getting the same feedback survey on cancelling and removing a product for WordPress.com products.
Should we be using the `skipRemovePlanSurvey` prop to skip the survey on removal? Or is filling the same survey at both cancel and removal useful?

![Screenshot 2025-06-26 at 5 13 04 PM](https://github.com/user-attachments/assets/c42b0cc0-16c7-4632-87bc-f90385fc5a96)

fditrapani replied:
Great question. I was thinking about it earlier. It can be worthwhile for us to gather this input in both spots for a few reasons:
- They can be happening at separate times.
- Different people can perform the action.
- they might leave even more candid feedback.

Since we'll eventually add a skip button, I think it'll be worth keeping. If anything, we might want to tweak the language to account for removal vs. cancelling. This label, in that case, should say "Why would you like to remove this upgrade?"

jboland88 reviewed (commented):
The functionality of cancelling first after a single confirmation screen is working as expected 👍 

I've noted a few things I noticed testing WordPress.com plans.

The one that I think would be nice to clean up here is skipping or altering the feedback survey on removal - perhaps so it's just one removal confirmation step. Particularly if we already ask the same questions after cancelling the product.

dzver committed:
Fix final removal button

dzver committed:
Change the title for confirm removal back.

dzver committed:
Add blue/red adjustment in Jetpack Cancel

dzver committed:
Requested by Josh

dzver committed:
Fix a bug in case the subscription is outside of refund period

The missing step would lead to displaying the default, which should
never be visible

dzver committed:
Move cancel survey post domain deletion

This change will send a new cancellation survey

dzver committed:
Update Keep subscription loading status

dzver commented:
I think I addressed all feedback but it's been lots of feedback so another round of testing is needed.

I also introduced the new flow for domains.

jboland88 commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -308,80 +235,48 @@ const CancelJetpackForm: React.FC< Props > = ( {
+				scary={ ! cancellationCompleted }
 			>
-				{ disabled ? cancellingText : cancelText }
+				{ cancellationCompleted ? translate( 'Submit' ) : translate( 'Cancel subscription' ) }
```
I think `cancellationCompleted` may not be getting passed through the `RemovePurchase` component, because I get the scary "Cancel Subscription" button when clicking "Remove Plan" from the purchase management screen

Although, having the removal action be "scary" might not be a bad thing.
I think it was just weird when for some reason it was showing on the survey prevuously 🤔 

![Screenshot 2025-06-27 at 5 43 52 PM](https://github.com/user-attachments/assets/ae75f281-a231-422c-81f2-9652bc6da32a)

jboland88 replied:
Nitpick: should I be able to click "Submit" on the survey if I have not chosen an answer?

![Screenshot 2025-06-27 at 5 59 25 PM](https://github.com/user-attachments/assets/e5db0d19-d713-42c5-b408-da3e01d5c882)

jboland88 commented on client/components/marketing-survey/cancel-jetpack-form/index.tsx:
```diff
@@ -308,80 +235,48 @@ const CancelJetpackForm: React.FC< Props > = ( {
+				scary={ ! cancellationCompleted }
 			>
-				{ disabled ? cancellingText : cancelText }
+				{ cancellationCompleted ? translate( 'Submit' ) : translate( 'Cancel subscription' ) }
```
Still a slight wording mismatch when removing an expired subscription between the copy in the modal and the copy of the button "Remove subscription" in the modal but the button says "cancel subscription" (Or "Submit" is the intended copy)

![Screenshot 2025-06-27 at 5 56 17 PM](https://github.com/user-attachments/assets/9dfc3699-d39c-4c35-b3d0-502098183001)

dzver replied:
I added the removal copy for the button and now get the proper text, thanks for finding this discrepancy!

jboland88 reviewed (commented):
I did not have any non-expired domains that were past the refund window to test with (and I'm not sure how to simulate that?) But the cancel and refund for domains worked as expected 👍 

Jetpack had a few oddities with the buttons showing in the cancellation modal (unexpected color or text) but otherwise tested well.

One small thing I noticed, after cancelling a purchase and getting a refund for Jetpack, I closed the survey and landed back on the purchases screen. The purchase I just cancelled + refunded was still showing in the list. Likely a state refresh needed there. Let's note that as a follow-up since it's out of scope for this PR:

![Screenshot 2025-06-27 at 6 08 05 PM](https://github.com/user-attachments/assets/1366bff8-102e-4856-a7b0-6825afe93ce5)

For WordPress.com purchases, I no longer am getting the survey on purchase removal, just a clear warning, which seems more correct IMO 👍 

![Screenshot 2025-06-27 at 6 31 09 PM](https://github.com/user-attachments/assets/e7b588b6-14e9-411f-bed6-cf0c93eecf3c)

fditrapani reviewed (approved):
Thanks for all the updates @dzver! this looks good on my end.

dzver committed:
Remove button in remove flow, Cancel in cancel flows
